llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -3872,6 +3872,10 @@ Instruction *InstCombinerImpl::visitOr(BinaryOperator &I) {
     }
   }
 
+  if (cast<PossiblyDisjointInst>(I).isDisjoint())
```

Comment:
It is ugly. But I don't know the best approach to treat `or disjoint` as `add nsw nuw`.

It is a phase ordering issue. We cannot assume that `ReassociatePass` will revert the canonicalization `add -> or disjoint`. BTW, I see this change achieves higher coverage in my benchmark (in `openexr/ImfTimeCode.cpp` and `z3/mpf.cpp`).

Reply:
We should probably have something like visitAddLike that is called from both visitAdd and visitOr where we can place folds we want to apply both to add and or disjoint?

Reply:
a less ugly way at least would be use `m_AddLike` matcher. But I agree with nikic.